Add loading images from prior sessions and sort by start time.

jkozol opened the pull request:
Images from prior sessions are now loaded at the start of a session. They will display in the blueprint page's list view sorted by start time. This PR should address issues #290 #269 #270 #271

jkozol commented:
I have found an issue with polling unfinished composes. I will push a fix tomorrow. This fix should also address #272

jgiardino commented on core/apiCalls.js:
```diff
@@ -176,3 +176,46 @@ export function fetchImageStatusApi(uuid) {
+      resolve(data.failed);
+    })
+    .catch(e => {
+      console.log(`Error fetching diff: ${e}`);
```
Should this message be more specific for each case?

jkozol replied:
Probably. Any opinions on how should we handle the error?

jgiardino replied:
I'm assuming "Error fetching diff" was added for /api/v0/blueprints/diff/.

I'd just have them identify the api call that's being made, like "Error fetching finished composes", etc...

Also, updates were made in #294 regarding the syntax for these. This syntax:
```console.log(`Error fetching diff: ${e}`);```
was updated to this:
`console.log('Error fetching diff', e);`

jgiardino commented on core/apiCalls.js:
```diff
@@ -176,3 +176,46 @@ export function fetchImageStatusApi(uuid) {
+    utils.apiFetch(constants.get_compose_queue)
+    .then(data => {
+      const queue = data.new.concat(data.run);
+      resolve(queue);
```
Does this get both composes that are pending plus the one that's in progress?

jkozol replied:
Yes.

larskarlitski replied:
I've always wondered why you're wrapping those calles in another `Promise`. `utils.apiFetch` already returns a promise, doesn't it?

jkozol replied:
You are correct. I will remove those extra promises from the compose API calls in this PR and will refactor the rest of the API calls in a future fix.

jgiardino commented on core/actions/composes.js:
```diff
@@ -0,0 +1,34 @@
+    blueprintName,
+    compose,
+  },
+});
```
Should we also add actions for handling errors for these?

jkozol replied:
I can add a error action alongside the fix I'm making.

jgiardino commented:
These updates look good! I can review again when you get the fix in that you mentioned, but I wanted to get your thoughts on something that I noticed. 

In the redux inspector, I'm noticing that we don't display contents on the Blueprints page until we've fetched blueprint contents. But as soon as we have the blueprint names, we're fetching both blueprint contents and composes? And it looks like we're getting information about composes before blueprint contents. Since we don't display contents on this page until blueprint contents are fetched, this results in taking a long time to load contents on the page. 

Is it possible to control the order that the state gets created, so that we wait until blueprint contents are loaded before getting the composes? I'm not sure this is a higher priority compared to other issues, but want to at least capture thoughts around how this could be solved later.

Can you describe how the `composes` array in the state gets it's data in the context of the Blueprints page? Maybe this is where I have a gap in my understanding of how redux-saga gets pulled in. 

BTW, reviewing this highlighted an issue with how I was handling the loading state for the Blueprints page, so I added issue #305.

jkozol commented:
@jgiardino 
In regards to the blueprint contents and composes loading, we should be able to check if the blueprint contents are loaded before calling the selector. If they aren't we can return an empty list and once they are the selector will get called to load the composes. I'll need to look more into this later, however, to make sure.

The way we load compose data is when a user opens the UI all of the composes will be immediately fetched from the API like we do with the basic blueprint info. When you go to a blueprint page a selector (getBlueprintComposes) is called which takes the compose state and selects the composes that correspond to the blueprint being viewed. These are then sorted by start time and the list is passed to the props of the blueprint page.

henrywang commented:
@jkozol, could you please fix the eslint issue? Thanks.
```
/home/travis/build/weldr/welder-web/core/apiCalls.js
  17:12  error  'err' is defined but never used            no-unused-vars
  27:12  error  'err' is defined but never used            no-unused-vars
  32:9   error  'page' is assigned a value but never used  no-unused-vars
  47:18  error  'e' is defined but never used              no-unused-vars
```

jkozol commented:
@henrywang Oops. Should be fixed now.

henrywang commented:
@jkozol, I got the "An Error Occurred" and "Internal error" when I load "http://localhost:9090/welder" first time. And web console reports "errorloadBlueprintsSaga". But it doesn't have same issue when I refresh this page. So it caused some end to end test cases failed. Could you please take a look? Thanks!
![screenshot_2018-07-19 image builder - cockpit](https://user-images.githubusercontent.com/672230/42919475-9b2a99a0-8b44-11e8-9f26-cf2fbc9b9a71.png)

jkozol commented:
@henrywang I am looking into the test failures. In regards to your errorloadBlueprintsSaga on first load, what environment are you experiencing this in? I am unable to replicate it.

henrywang commented:
@jkozol, here's my review workflow:
1. checkout PR branch.
2. node run build
3. sudo docker run -d --name web --restart=always -v bdcs-socket:/run/weldr -v `pwd`/public:/usr/share/cockpit/welder --network host welder/web-lorax:latest
4. open http://localhost:9090/welder

jgiardino commented:
These updates look good!

The only issue that I notice is there was a failed compose already in my vm. When I look at it in the list of images, the Date Created is showing today, but this is one that I had started last week. Maybe timestamps don't really reflect the start time? Maybe this was really an issue with my vm?  If there's a way to check using the cli what the timestamp is, I would, but I'm not sure there is a command for that. Let me know if you have any thoughts on this one. 

As for handling when we get composes (i.e. immediately on page load vs on user action to view composes), I don't think we need to make any changes in this PR. This would be a good topic for discussion next week, and is probably a larger topic that potentially requires other changes than just how we get composes.

bcl commented:
Timestamps are currently the last time the STATUS file of a compose was updated. This could be changed, and/or we could also add a start time by looking at the timestamp from one of the other metadata files like final-kickstart.ks

jkozol commented:
@henrywang I added explicit actions to trigger loading blueprints and composes and modified the sagas a little bit. Now, only the import content sanity test fails. Hopefully I can fix that issue during the hackathon.
